contents/blog/hog-story.md

```diff
@@ -0,0 +1,206 @@
+
+This is a lot of power to hand to our users, and not something they could easily find in competitors. But was it worth the effort to realize this vision?
+
+Ours is a culture that doesn't fear hard problems, but wasting energy isn't a virtue either.
+
+An obvious and reasonable objection that came up in one of Marius’s [RFC’s](handbook): it would add layers of complexity we would then be responsible for understanding, forever. What happens when it fails, as computing systems inevitably do? How would we debug problems? 
+
+But there were also rewards for this complexity. For one thing, a fully realized HogQL would enable us to nuke a huge pile of technical debt.
+
+A central feature of PostHog product analytics is *insights*: these are the various graphs and tables you can customize and attach to dashboards. Each insight was a tangle of duplicated Python, and working with them was a pain.
+
+The proposed HogQL was powerful enough to both use ourselves – to replace the internals of these [insights]() – while also being something our customers could use to perfectly tailor their own reporting.
+
+HogQL even lowered the costs of building new features, like [Data Warehouse]().
+
+These benefits overcame the uncertainty of the process. It was worth trying.
+
+And it worked: today HogQL is used everywhere in PostHog, and you can use it yourself just as envisioned.
+
+It was the first step to making PostHog into a *platform* where developers could host their own business logic for understanding user behavior in their products.
+
+# Seeing The Matrix with Hog
+
+Queries are necessary but not sufficient for such a platform.
+
+You also need to be able to customize logic and interactions with other systems. This would address a problem nearly every PostHog customer encounters transforming and exporting data.
+
+Deeper in *Crafting Interpreters,* Marius knew, lay the key.
+
+The naive solution to giving developers customizable behavior would be letting them write JavaScript. But there are so many issues with running arbitrary JavaScript at scale.
+
+" We had multiple problems with our JavaScript-based dynamic code execution engine. Security was the major one that just killed it at the end. But the other big problem was that JavaScript is kind of unpredictable. You start a script, and it can just leak promises everywhere," Marius told me.
+
+This just wouldn’t work for code that could be triggered by any individual user event. “ There's some callback that's going to resolve like two hours later and you have no idea that it's out there, eating memory. You do that for tens of thousands of events every minute… That's going to crash.”
+
+This wasn't speculative, either. Our own internal, JavaScript-based plugin engine had these problems all the time. It was impossible to expose the same power to external developers and also keep the product online.
+
+JavaScript was limiting the scope of our ambition.
+
+# What's a bytecode, anyway?
```

Comment:
Also I think you could halve this section to help people get to the next one sooner, feel like you repeat the same points a few times here. For example:

- "The HogVM is implemented in both Python and TypeScript. Someday, if we need even more performance, it may be built again in Rust." isn't super relevant.
- "Because we control the language and the environment where it runs, we can do some interesting voodoo:" is extra
- It feels like the "completely control" and "customize" components get repeated multiple times.